linkerd/core/src/main/scala/io/buoyant/linkerd/Linker.scala

```diff
@@ -21,46 +20,66 @@ trait Linker {
 
 object Linker {
 
-  def parse(config: String, configInitializers: Seq[ConfigInitializer]): LinkerConfig = {
-    val mapper = Parser.objectMapper(config)
-    // Subtypes must not conflict
-    configInitializers.groupBy(_.configId).collect {
-      case (id, cis) if cis.size > 1 =>
-        throw ConflictingSubtypes(cis(0).namedType, cis(1).namedType)
+  private def ensureUniqueKinds(inits: Seq[ConfigInitializer]): Unit =
+    inits.groupBy(_.configId).foreach {
+      case (_, Seq(a, b, _*)) => throw ConflictingSubtypes(a.namedType, b.namedType)
+      case _ =>
     }
-    for (ci <- configInitializers) ci.registerSubtypes(mapper)
-    // TODO: Store the LinkerConfig so that it can be serialized out later
-    mapper.readValue[LinkerConfig](config)
-  }
 
-  def load(config: String, configInitializers: Seq[ConfigInitializer]): Linker = {
-    parse(config, configInitializers).mk
-  }
+  private[linkerd] case class Initializers(
+    protocol: Seq[ProtocolInitializer] = Nil,
+    namer: Seq[NamerInitializer] = Nil,
+    interpreter: Seq[InterpreterInitializer] = Nil,
+    tlsClient: Seq[TlsClientInitializer] = Nil,
+    tracer: Seq[TracerInitializer] = Nil
+  ) {
+    def iter: Iterable[Seq[ConfigInitializer]] =
+      Seq(protocol, namer, interpreter, tlsClient, tracer)
+
+    def parse(config: String): LinkerConfig =
+      Linker.parse(config, this)
 
-  def parse(config: String): LinkerConfig = {
-    val protocols = LoadService[ProtocolInitializer]
-    val namers = LoadService[NamerInitializer]
-    val clientTls = LoadService[TlsClientInitializer]
-    val tracers = LoadService[TracerInitializer]
-    parse(config, protocols ++ namers ++ clientTls ++ tracers)
+    def load(config: String): Linker =
+      Linker.load(config, this)
   }
 
-  def load(config: String): Linker = {
-    parse(config).mk
+  private[linkerd] lazy val LoadedInitializers = Initializers(
+    LoadService[ProtocolInitializer],
+    LoadService[NamerInitializer],
+    LoadService[InterpreterInitializer] :+ DefaultInterpreterInitializer,
+    LoadService[TlsClientInitializer],
+    LoadService[TracerInitializer]
+  )
+
+  private[linkerd] def parse(
+    config: String,
+    inits: Initializers = LoadedInitializers
+  ): LinkerConfig = {
+    val mapper = Parser.objectMapper(config)
+    for (kinds <- inits.iter) {
+      ensureUniqueKinds(kinds)
+      kinds.foreach(_.registerSubtypes(mapper))
```

Comment:
A bit dissonant to use for and foreach right next to each other.  Maybe use for here to be consistent?